File: scala-simple-http-server/src/main/scala/RequestParser.scala
import java.io.{BufferedReader, InputStream, InputStreamReader}

object RequestParser {

  def fromInputStream(in: InputStream): Option[Request] = {
    val reader = new BufferedReader(new InputStreamReader(in))
    val requestLine = reader.readLine()

    Option(requestLine).map { line => // requestLineはnullかも知れないのでOptionでくるむ
      val Array(method, targetPath, httpVersion) = line.split("\\s") // 変数x, y, zにArrayの中身を束縛。Arrayの長さが3ではない場合は例外をスローする
      Request(method, targetPath, httpVersion)
    }
  }
}

case class Request(
  method: String,
  targetPath: String,
  httpVersion: String)

File: java-simple-http-server/src/main/java/server/RequestParser.java
package server;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RequestParser {

    public static Pattern requestLinePattern = Pattern.compile("(?<method>.*) (?<path>.*?) (?<version>.*?)");

    /**
     * InputStreamからHTTPリクエストをパースし、Requestを生成する。
     * InvalidなHTTPリクエストの場合はnullを返す。
     */
    public Request fromInputStream(InputStream in) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(in));
        String requestLine = reader.readLine();

        if (requestLine == null) return null;

        Matcher matcher = requestLinePattern.matcher(requestLine);

        if (!matcher.find()) return null;

        String method      = matcher.group("method");
        String targetPath  = matcher.group("path");
        String httpVersion = matcher.group("version");

        return new Request(method, targetPath, httpVersion);
    }
}
